clang/lib/Driver/Driver.cpp

```diff
@@ -4482,55 +4615,83 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
       Action *Current = C.MakeAction<InputAction>(*InputArg, T);
       UnbundlerInputs.push_back(Current);
     };
-    for (const auto *A : Args.filtered(options::OPT_foffload_static_lib_EQ))
-      addUnbundlerInput(types::TY_Archive, A);
-    for (const auto *A :
-        Args.filtered(options::OPT_foffload_whole_static_lib_EQ))
-      addUnbundlerInput(types::TY_WholeArchive, A);
+    // Go through all of the args, and create a Linker specific argument list.
+    // This is fed into the partial link step.
+    SmallVector<const char *, 16> LinkArgs;
+    getLinkerArgs(C, Args, LinkArgs);
+    bool IsWholeArchive = false;
+    for (std::string MA : LinkArgs) {
+      if (isStaticArchiveFile(MA)) {
+        addUnbundlerInput(IsWholeArchive ? types::TY_WholeArchive
+                                         : types::TY_Archive,
+                          MA.c_str());
+        continue;
+      }
+      if (optionMatches("-no-whole-archive", MA)) {
+        IsWholeArchive = false;
+        continue;
+      }
+      if (optionMatches("-whole-archive", MA)) {
+        IsWholeArchive = true;
+        continue;
+      }
+      if (isObjectFile(MA)) {
+        // Add any objects to the unbundler step.  These objects are passed
+        // directly to the linker, so the driver does not know about them.
+        // FIXME - Better process objects passed to the linker process them
+        // like regular objects that need to be potentially unbundled.
+        addUnbundlerInput(types::TY_Object, MA.c_str());
+        continue;
+      }
+    }
+
     if (!UnbundlerInputs.empty()) {
-      Action *Current = C.MakeAction<InputAction>(*LastArg, types::TY_Archive);
-      OffloadBuilder.addHostDependenceToUnbundlingAction(Current,
-          UnbundlerInputs, LastArg);
+      Action *PartialLink =
+          C.MakeAction<PartialLinkJobAction>(UnbundlerInputs, types::TY_Object);
+      Action *Current = C.MakeAction<InputAction>(*LastArg, types::TY_Object);
+      ActionList AL;
+      AL.push_back(PartialLink);
+      OffloadBuilder.addHostDependenceToUnbundlingAction(Current, AL, LastArg);
       Current = OffloadBuilder.addDeviceDependencesToHostAction(Current,
           LastArg, phases::Link, PL.back(), PL);
     }
   }
   const llvm::opt::OptTable &Opts = getOpts();
-  auto unbundleStaticLib = [&](types::ID T, const Arg *A) {
-    Arg *InputArg = MakeInputArg(Args, Opts, A->getValue());
+  auto unbundleStaticLib = [&](types::ID T, const std::string &A) {
+    Arg *InputArg = MakeInputArg(Args, Opts, Args.MakeArgString(A));
     Action *Current = C.MakeAction<InputAction>(*InputArg, T);
     OffloadBuilder.addHostDependenceToDeviceActions(Current, InputArg, Args);
     OffloadBuilder.addDeviceDependencesToHostAction(
         Current, InputArg, phases::Link, PL.back(), PL);
   };
-  for (const auto *A : Args.filtered(options::OPT_foffload_static_lib_EQ)) {
+  SmallVector<const char *, 16> LinkArgs;
+  getLinkerArgs(C, Args, LinkArgs);
```

Comment:
This essentially duplicates 4627-4628. Could a single routine call within the function suffice?

Reply:
Yup, will simplify!